From 4a243b378743a2d3bd0f579097cbf1c405fe8156 Mon Sep 17 00:00:00 2001
From: James Elliott <jjellio@sandia.gov>
Date: Tue, 27 Apr 2021 19:57:57 -0600
Subject: [PATCH] add HIP mirroring how Cuda works.

CHECK:
If ATDM enables CUDA it also enables TPL_CUDA and TPL_CUSPARSE
Need to see if we have TPLs for those, and if not add them
TPL_HIP and TPL_ROCSparse
---
 cmake/std/atdm/ATDMDevEnvSettings.cmake | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/cmake/std/atdm/ATDMDevEnvSettings.cmake b/cmake/std/atdm/ATDMDevEnvSettings.cmake
index 3735bd10b2d..c5e87b649f5 100644
--- a/cmake/std/atdm/ATDMDevEnvSettings.cmake
+++ b/cmake/std/atdm/ATDMDevEnvSettings.cmake
@@ -44,6 +44,7 @@ ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(PARALLEL_LINK_JOBS_LIMIT "")
 ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(Trilinos_LINK_SEARCH_START_STATIC OFF)
 ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(USE_OPENMP OFF)
 ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(USE_PTHREADS OFF)
+ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(USE_HIP OFF)
 ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(USE_CUDA OFF)
 ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(CUDA_RDC OFF)
 ATDM_SET_ATDM_VAR_FROM_ENV_AND_DEFAULT(ADDRESS_SANITIZER OFF)
@@ -66,6 +67,9 @@ IF (ATDM_USE_OPENMP)
 ELSEIF (ATDM_USE_CUDA)
   SET(ATDM_NODE_TYPE CUDA)
   SET(ATDM_INST_SERIAL ON)
+ELSEIF (ATDM_USE_HIP)
+  SET(ATDM_NODE_TYPE HIP)
+  SET(ATDM_INST_SERIAL ON)
 ELSE()
   SET(ATDM_NODE_TYPE SERIAL)
   SET(ATDM_INST_SERIAL ON)
@@ -302,6 +306,7 @@ ATDM_SET_CACHE(Kokkos_ENABLE_CUDA_UVM "${ATDM_USE_CUDA}" CACHE BOOL)
 ATDM_SET_CACHE(Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE "$ENV{ATDM_CONFIG_CUDA_RDC}"
   CACHE BOOL)
 ATDM_SET_CACHE(Kokkos_ENABLE_SERIAL "${ATDM_Kokkos_ENABLE_SERIAL}" CACHE BOOL)
+ATDM_SET_CACHE(Kokkos_ENABLE_HIP "${ATDM_USE_HIP}" CACHE BOOL)
 ATDM_SET_CACHE(Kokkos_ENABLE_CUDA "${ATDM_USE_CUDA}" CACHE BOOL)
 ATDM_SET_CACHE(Kokkos_ENABLE_CXX11_DISPATCH_LAMBDA ON CACHE BOOL)
 ATDM_SET_CACHE(Kokkos_ENABLE_CUDA_LAMBDA "${ATDM_USE_CUDA}" CACHE BOOL)
@@ -324,6 +329,7 @@ ATDM_SET_CACHE(Panzer_FADTYPE "Sacado::Fad::DFad<RealType>" CACHE STRING)
 ATDM_SET_CACHE(Phalanx_KOKKOS_DEVICE_TYPE "${ATDM_NODE_TYPE}" CACHE STRING)
 ATDM_SET_CACHE(Phalanx_SHOW_DEPRECATED_WARNINGS OFF CACHE BOOL)
 ATDM_SET_CACHE(DAKOTA_ENABLE_TESTS OFF CACHE BOOL)
+ATDM_SET_CACHE(Tpetra_INST_HIP "${ATDM_USE_HIP}" CACHE BOOL)
 ATDM_SET_CACHE(Tpetra_INST_CUDA "${ATDM_USE_CUDA}" CACHE BOOL)
 ATDM_SET_CACHE(Tpetra_INST_SERIAL "${ATDM_INST_SERIAL}" CACHE BOOL)
 ATDM_SET_CACHE(Tpetra_INST_INT_INT OFF CACHE BOOL)
@@ -331,7 +337,7 @@ ATDM_SET_CACHE(Xpetra_ENABLE_Epetra OFF CACHE BOOL)
 ATDM_SET_CACHE(Amesos2_ENABLE_Epetra OFF CACHE BOOL)
 ATDM_SET_CACHE(MueLu_ENABLE_Epetra OFF CACHE BOOL)
 ATDM_SET_CACHE(Piro_ENABLE_MueLu OFF CACHE BOOL)
-IF (ATDM_USE_CUDA)
+IF ((ATDM_USE_CUDA) OR (ATDM_USE_HIP))
   ATDM_SET_CACHE(Sacado_ENABLE_HIERARCHICAL_DFAD ON CACHE BOOL)
 ENDIF()
 ATDM_SET_CACHE(DART_TESTING_TIMEOUT 600 CACHE STRING)
@@ -370,6 +376,10 @@ ELSE()
   SET(ATDM_TPL_LIB_EXT "a")
 ENDIF()
 
+# HIP? I don't think we have a TPL for it yet
+ATDM_SET_ENABLE(TPL_ENABLE_HIP "${ATDM_USE_HIP}")
+ATDM_SET_ENABLE(TPL_ENABLE_ROCSPARSE "${ATDM_USE_HIP}")
+
 # CUDA
 ATDM_SET_ENABLE(TPL_ENABLE_CUDA "${ATDM_USE_CUDA}")
 
